### PR TITLE
647 case details

### LINKED ISSUE #647: RVL-838 - Show Index Case details on Manage Plans screen for reactive intervention
This new information needs to be shown on Manage Plans page for FI plans:

### Change to be made:

1. Show Case Number and sdd a collapsible section labelled “Case Details” under “Case Number” with fields  (need to confirm if all of these fields pulling from BIOPHICS into Reveal)
    - Case Information
        - Case Classification
        - Date of Diagnosis
        - Species
        - Notification Date
        - Investigation Date
        - House Number
        - Patient Name
        - Surname
        - Age
    - Foci Information
        - Foci of Residence
            - ID
            - Classification
            - Name
        - Foci of Infection (Blank if same as above)
            - ID
            - Classification
            - Name

### Business Logic:

- One index case should return
- All fields are read only
- The section should be shown by default with the ability to collapse
- The section only displays if the Focus Investigation Reason = Case Triggered

More info: https://smartregister.atlassian.net/browse/RVL-838

### PR DESCRIPTION
**Pull request by p-netm**
closes #647.
adds a `CaseDetails` component that is embedded to the manage plans page that show the data as requested in the tagged issue.

### CODE REVIEW

**Comment by p-netm** [comment]
@moshthepitt Just a few things remaining:
- some tests for 2 utility functions,
- And getting the styling for the text right.

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+/** CaseDetails component */
+export const CaseDetails: React.FC<CaseDetailsProps> = (props = defaultCaseDetailsProps) => {
+  const { eventId, service, event, fetchEventsCreator } = props;
+  const [isCollapsed, setCollapse] = React.useState<boolean>(false);
+  const toggleCollapse = () => setCollapse(!isCollapsed);
+
+  if (eventId === null) {
+    // if eventId is null return early
+    return <Loading />;
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: If eventId is false-y, should we show a loading animation, should we just display an error message?

Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: yeah returning an error better fits this situation

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+/** CaseDetails component */
+export const CaseDetails: React.FC<CaseDetailsProps> = (props = defaultCaseDetailsProps) => {
+  const { eventId, service, event, fetchEventsCreator } = props;
+  const [isCollapsed, setCollapse] = React.useState<boolean>(true);
+  const toggleCollapse = () => setCollapse(!isCollapsed);
+
+  if (eventId === null) {
+    // if eventId is null return early
+    displayError(new Error(FAILED_TO_GET_EVENT_ID));
+    return null;
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Do we have a test that covers this?

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+  // make a call to the events endpoint and get event
+  useEffect(() => {
+    loadEvent(eventId, service, fetchEventsCreator).catch((err: Error) =>
+      growl(err.message, { type: toast.TYPE.ERROR })
+    );
+  }, []);
+
+  if (!event) {
+    return <Loading />;
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Do we have a test for this case?

Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: yes, for this one we have.

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx (hunk 2 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Ill add one for this.

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/tests/index.test.tsx [hunk 4]
```diff
@@ -49,6 +49,24 @@ describe('src/containers/pages/interventionPlan/updateplan/caseDetials', () => {
     expect(wrapper.find('Ripple').length).toEqual(0);
     expect(wrapper.text()).toMatchSnapshot(`Case Details`);
   });
+  it('renders nothing when eventId is null', async () => {
```
Comment (moshthepitt) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/tests/index.test.tsx]: @p-netm does this test that displayError was called with the expected error message?

File: src/configs/lang.ts [hunk 5]
```diff
@@ -411,3 +411,20 @@ export const IRS_GREY_THRESHOLD = translate('IRS_GREY_THRESHOLD', 'Grey');
...
+export const CASE_DETAILS = translate('CASE_DETAILS', 'Case Details');
+export const CASE_INFORMATION = translate('CASE_INFORMATION', 'Case Information');
+export const AGE = translate('AGE', 'Age');
+export const SURNAME = translate('SURNAME', 'Surname');
+export const PATIENT_NAME = translate('PATIENT_NAME', 'Patient Name');
+export const HOUSE_NUMBER = translate('HOUSE_NUMBER', 'House Number');
+export const INVESTIGATION_DATE = translate('INVESTIGATION_DATE', 'Investigation Date');
+export const SPECIES = translate('SPECIES', 'Species');
+export const DIAGNOSIS_DATE = translate('DIAGNOSIS_DATE', 'Date of Diagnosis');
+export const ID = translate('ID', 'Id');
```
Comment (cKellyDesign) [inline src/configs/lang.ts]: I thought we already had one for ID (which, in TH was "translated" to ID). Is it important that the `d` is lowercase? Should it be uppercase b/c it's an abbreviation and not [Freud's Id](https://en.wikipedia.org/wiki/Id,_ego_and_super-ego#Id)?

File: src/constants.tsx [hunk 6]
```diff
@@ -88,6 +88,7 @@ export const OPENSRP_ORG_PRACTITIONER_ENDPOINT = 'organization/practitioner';
 export const OPENSRP_DEL_PRACTITIONER_ROLE_ENDPOINT = 'practitionerRole/deleteByPractitioner';
 export const OPENSRP_ADD_PRACTITIONER_ROLE_ENDPOINT = 'practitionerRole/add';
 export const OPENSRP_USERS_ENDPOINT = 'user';
+export const OPENSRP_FIND_EVENTS_ENDPOINT = 'event/findById';
```
Comment (cKellyDesign) [inline src/constants.tsx]: Is this in the `.env.sample`?

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  fetchEventsCreator: fetchEvents,
+  service: OpenSRPService,
+};
+
+/** async function to load a single event
+ * @param {UUID} eventId - id of the event to get
+ * @param {OpenSRPService} servant - the opensrp service
+ * @param {ActionCreator<FetchEventsAction>} actionCreator - the actionCreator
+ */
+const loadEvent = async (
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Do we need `async` here if we're not `await`ing?

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    id: eventId,
+  };
+  service
+    .list(filterParams)
+    .then((response: RawEvent) => actionCreator([response]))
+    .catch((err: Error) => displayError(err));
+};
+
+/** CaseDetails component */
+export const CaseDetails: React.FC<CaseDetailsProps> = (props = defaultCaseDetailsProps) => {
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Should `props` have a type here? Or is that adequately covered by `React.FC<CaseDetailsProps>`?

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    displayError(new Error(FAILED_TO_GET_EVENT_ID));
+    return null;
+  }
+
+  // make a call to the events endpoint and get event
+  useEffect(() => {
+    loadEvent(eventId, service, fetchEventsCreator).catch((err: Error) =>
+      growl(err.message, { type: toast.TYPE.ERROR })
+    );
+  }, []);
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Should this effect be made when any of the props are updated?

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+  if (!event) {
+    return <Loading />;
+  }
+  const translatedEvent = translateEvent(event);
+  const { caseInformation, fociInformation, caseNumber, fociInformationTitle } = translatedEvent;
+
+  const mapObjectToElement = (obj: FlexObject) =>
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Nit-pick: can we change this to `mapObjectToTr` to be more specific?

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        </Card>
+      </div>
+    </>
+  );
+};
+
+/** describes object returned by mapState */
+type MapStateToProps = Pick<CaseDetailsProps, 'event'>;
+/** describes mapDispatch */
+type DispatchToProps = Pick<CaseDetailsProps, 'fetchEventsCreator'>;
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Are these specific to this component? If not, should we move them to the `types` file?

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+/**
+ *
+ * @param {Partial<Store>} state = the store slice
+ * @param {CaseDetailsProps} ownProps - the props
+ */
+const mapStateToProps = (state: Partial<Store>, ownProps: CaseDetailsProps): MapStateToProps => {
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Shall we give this the type `MapStateToProps`?

File: src/containers/pages/InterventionPlan/UpdatePlan/index.tsx [hunk 13]
```diff
@@ -103,6 +113,12 @@ const UpdatePlan = (props: RouteComponentProps<RouteParams> & UpdatePlanProps) =
         <Col md={8}>
           <PlanForm {...planFormProps} />
         </Col>
+        <Col md={4}>
+          {/* Only show case details if plan is reactive */}
+          {planIsReactive(plan) && (
+            <ConnectedCaseDetails {...{ ...caseDetailsProps, eventId: getEventId(plan) }} />
+          )}
+        </Col>
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/index.tsx]: :heart:

File: src/containers/pages/InterventionPlan/UpdatePlan/utils.tsx [hunk 14]
```diff
@@ -0,0 +1,32 @@
+import { FAILED_TO_GET_EVENT_ID } from '../../../../configs/lang';
+import { PlanDefinition, UseContext } from '../../../../configs/settings';
+import { displayError } from '../../../../helpers/errors';
+
+export const getEventId = (aPlan: PlanDefinition | null): string | null => {
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/utils.tsx]: Doc strings please

File: src/containers/pages/InterventionPlan/UpdatePlan/utils.tsx [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+  });
+
+  if (contextOfInterest.length !== 1) {
+    displayError(new Error(FAILED_TO_GET_EVENT_ID));
+    return null;
+  }
+  return contextOfInterest[0].valueCodableConcept;
+};
+
+export const planIsReactive = (aPlan: PlanDefinition | null): boolean => {
```
Comment (cKellyDesign) [inline src/containers/pages/InterventionPlan/UpdatePlan/utils.tsx]: Doc strings please

File: src/store/ducks/opensrp/events/index.tsx [hunk 16]
```diff
@@ -0,0 +1,106 @@
+/** Events redux module */
+import { Dictionary, get, keyBy } from 'lodash';
+import { Store } from 'redux';
+import { AnyAction } from 'redux';
+import SeamlessImmutable from 'seamless-immutable';
+import { Event, extractEvents, RawEvent } from './utils';
```
Comment (cKellyDesign) [inline src/store/ducks/opensrp/events/index.tsx]: @p-netm @moshthepitt - where do you think is the best place for these type definitions?

File: src/store/ducks/opensrp/events/utils.ts [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+    case_classification: string;
+  };
+  version: number;
+  teamId: UUID;
+  _id: UUID;
+  _rev: string;
+}
+
+/** describes the foci information property */
+interface FociInformation {
```
Comment (cKellyDesign) [inline src/store/ducks/opensrp/events/utils.ts]: Let's export this as well 🙏

File: src/store/ducks/opensrp/events/utils.ts [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+  id: UUID;
+  baseEntityId: UUID;
+  caseNumber: string;
+  caseInformation: CaseInformation;
+  fociInformation: FociInformation;
+  fociInformationTitle: string;
+}
+
+/** lookup table to decide the title of foci information based on events flag details */
+const FociTitleLookup: Dictionary<string> = {
```
Comment (cKellyDesign) [inline src/store/ducks/opensrp/events/utils.ts]: Should this be exported? Let's also change to `fociTitleLookup` b/c it's not a class or a type.

File: src/store/ducks/opensrp/events/utils.ts [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+};
+
+/** describes the form an event will take after translating its properties */
+interface TranslatedEvent {
+  baseEntityId: UUID;
+  caseInformation: Dictionary<string>;
+  caseNumber: string;
+  fociInformation: Dictionary<string>;
+  fociInformationTitle: string;
+}
```
Comment (cKellyDesign) [inline src/store/ducks/opensrp/events/utils.ts]: What's the difference between `Event` and `TranslatedEvent`? Can one extend another?

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/tests/index.test.tsx (hunk 4 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/tests/index.test.tsx]: add that.

File: src/configs/lang.ts (hunk 5 above)
Comment (p-netm) [reply src/configs/lang.ts]: @cKellyDesign It actually needs to be ID, but i could not find it in `config/lang.ts`

Comment (p-netm) [reply src/configs/lang.ts]: Scratch that, found it.

File: src/constants.tsx (hunk 6 above)
Comment (p-netm) [reply src/constants.tsx]: @cKellyDesign Doesn't seem like we have been placing the endpoints in env.sample.

Comment (p-netm) [reply src/constants.tsx]: If there should be there then maybe i can create a ticket to move them there

Comment (moshthepitt) [reply src/constants.tsx]: I think @cKellyDesign agrees with you that we don't need to do that :)

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx (hunk 7 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: will remove that

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx (hunk 8 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: yes, `React.FC<props>` is more strict that inline annotating the type in the parenthesis

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx (hunk 9 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: yes., I will change that

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx (hunk 10 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: ok

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx (hunk 11 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: Yes this are specific to CaseDetails

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx (hunk 12 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: @cKellyDesign  Not sure i understand, there is a chance am intoxicated, however I will say that `MapStateToProps` describes an object that mapState function returns.

File: src/containers/pages/InterventionPlan/UpdatePlan/index.tsx (hunk 13 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/index.tsx]: :+1:

File: src/containers/pages/InterventionPlan/UpdatePlan/utils.tsx (hunk 14 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/utils.tsx]: ok, I can do that.

File: src/containers/pages/InterventionPlan/UpdatePlan/utils.tsx (hunk 15 above)
Comment (p-netm) [reply src/containers/pages/InterventionPlan/UpdatePlan/utils.tsx]: done

File: src/store/ducks/opensrp/events/index.tsx (hunk 16 above)
Comment (p-netm) [reply src/store/ducks/opensrp/events/index.tsx]: I can create an Issue to discuss and implement.

File: src/store/ducks/opensrp/events/utils.ts (hunk 17 above)
Comment (p-netm) [reply src/store/ducks/opensrp/events/utils.ts]: ok

File: src/store/ducks/opensrp/events/utils.ts (hunk 18 above)
Comment (p-netm) [reply src/store/ducks/opensrp/events/utils.ts]: ok

File: src/store/ducks/opensrp/events/utils.ts (hunk 19 above)
Comment (p-netm) [reply src/store/ducks/opensrp/events/utils.ts]: @moshthepitt @cKellyDesign The event is store in store and extracted from the rawEvent, howver the translatedEvent will be used during display , such that i can iterate over its keys and values display them without needing to worry about localizing each property name at render time

File: src/constants.tsx (hunk 6 above)
Comment (cKellyDesign) [reply src/constants.tsx]: oh sorry, for some reason i thought this was in `env.ts`, my bad!

File: src/store/ducks/opensrp/events/utils.ts (hunk 19 above)
Comment (cKellyDesign) [reply src/store/ducks/opensrp/events/utils.ts]: @p-netm ok cool, maybe we can just add `for display` somewhere in the doc-string?

File: src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx (hunk 8 above)
Comment (cKellyDesign) [reply src/containers/pages/InterventionPlan/UpdatePlan/CaseDetails/index.tsx]: love it

File: src/store/ducks/opensrp/events/utils.ts (hunk 19 above)
Comment (p-netm) [reply src/store/ducks/opensrp/events/utils.ts]: ok

Comment (cKellyDesign) [reply src/store/ducks/opensrp/events/utils.ts]: @p-netm ok cool, maybe we can just add `for display` somewhere in the doc-string?

File: src/store/ducks/opensrp/events/index.tsx (hunk 16 above)
Comment (p-netm) [reply src/store/ducks/opensrp/events/index.tsx]: #736

**Pull request merged**